src/coreclr/nativeaot/Runtime/unix/UnwindHelpers.cpp

```diff
@@ -761,8 +761,85 @@ void Registers_REGDISPLAY::setVectorRegister(int num, libunwind::v128 value)
 
 #endif // TARGET_ARM64
 
+// the cache relies on "endless" 64bit version counter
+#if TARGET_64BIT
+struct ProcInfoCacheEntry
+{
+    PCODE pc;
+    size_t version;
+    unw_proc_info_t procInfo;
+};
+
+// we use static array with 512 entries as a cache
+// that is about 45Kb  (assuming unw_proc_info_t is 9 * 8 bytes) and what we can reasonably afford.
+static const int CACHE_BITS = 9;
+static ProcInfoCacheEntry cache[1 << CACHE_BITS];
```

Comment:
Any numbers for how much does this cache help?

We had several variants of caches like this over the years and they never worked well.

Reply:
Difficulties with caching this kind of operations is that the cache must be faster than what is cached. What we have is a relatively cheap operation that is performed frequently over the same data. 
It is only "relatively" cheap though - libunwind would do binary search among tens of thousands of records, with some indirections and some parsing, may have to take locks to ensure that records are not changing due to unloading. That adds up.

Scenarios that are sensitive to GC pauses should see improvements. I'll get perf numbers for the Stage2 scenario.

Reply:
I had some preliminary measurements, but wanted to be sure to compare with the newest build.

So here the impact on a Stage2 benchmark - an ASP.Net app that has some allocations.
(Linux, x64, 28 cores)

=== before
```
| Requests/sec           | 71,516    |
| Requests               | 1,078,167 |
| Mean latency (ms)      | 4.35      |
| Max latency (ms)       | 35.06     |
```

=== after
```
| Requests/sec           | 80,962    |
| Requests               | 1,221,769 |
| Mean latency (ms)      | 3.71      |
| Max latency (ms)       | 25.78     |
```

Reply:
Here where we spend time in the stack walk:
(after the caching change the cost of getInfoFromDwarfSection no longer matters)  

=== before
![image](https://github.com/dotnet/runtime/assets/8218165/7baa0ea6-03ba-419d-ab4d-073fbf67ba5b)

=== after
![image](https://github.com/dotnet/runtime/assets/8218165/fb769323-351d-452e-890d-a80491329668)

Reply:
On Windows, the improvements are less impressive. Unlike ~13% improvement on Linux, I see only about ~1% on Windows. 

The reasons could be that searching for method info on Windows is simpler. There is just a binary search, although in a pretty large array.  

Also for the Stage2 benchmark not all problems could be GC related. I see a 0.5-1 sec max pause, for example. It is unlikely to be caused by GC - not with this tiny heap size. I am not sure if this mega-pause is something new. I do not recall seeing it before.

=== before
```
| Requests/sec           | 61,163  |
| Requests               | 922,488 |
| Mean latency (ms)      | 7.55    |
| Max latency (ms)       | 537.89  |
```

=== after change
```
| Requests/sec           | 62,111  |
| Requests               | 937,148 |
| Mean latency (ms)      | 7.08    |
| Max latency (ms)       | 534.79  |
```

Reply:
Perhaps we should drop caching changes on Windows from this PR until we have better data confirming the gain. 

The other changes, like not unwinding SSE registers for GC stackwalks are less controversial, even if we do not see much impact.

Reply:
> Perhaps we should drop caching changes on Windows from this PR until we have better data confirming the gain.

Yes, it would make sense.

Also, since the binary search can be very fast on Windows, is the problem on non-Windows in the data structure design or just in a poor implementation of the binary search (e.g. nobody cared about the perf of this binary search before)?

> not unwinding SSE registers for GC stackwalks are less controversial,

Agree, these changes look good to me.

Reply:
> is the problem on non-Windows in the data structure design or just in a poor implementation of the binary search 

In Windows case the method info appears to be present as-is and there is also an index of pointers that we search through. That is faster, although it may still take time if we do it repeatedly. The number of active frames could be vastly smaller than the overall number of entries. (i.e. most frames are rarely active, long stacks tend to be recursive or a bunch of same async state machine frames).  
I do not know where the design comes from but it might be a deliberate tradeoff - it may take more space in the file and in memory, but since the data can be used as-is and there is an index, finding a method info for a given PC is simpler.

In DWARF case there is some parsing and fetching the data from which the result is composed. It looks like there could be different formats (compact and regular). The datastructure that is composed is not present in the file as-is. I think this scheme was designed more for one-off unwind scenarios like debugging or exception handling. Those are more random, but less time sensitive than GC stackwalks.

Reply:
I've removed caching on Windows.

Reply:
> > not unwinding SSE registers for GC stackwalks are less controversial,

> Agree, these changes look good to me.

Maybe move these non-controversial changes to a separate PR so this one just about the cache?